File: typescript/packages/core/src/types/schemes/upto.ts
/**
 * Up-To Scheme Types
 *
 * The `upto` scheme authorizes transfer of up to a maximum amount,
 * enabling usage-based billing where the final settlement amount
 * is determined by actual usage.
 *
 * @example
 * ```typescript
 * // Client authorizes up to $1.00
 * const requirements: UptoPaymentRequirements = {
 *   scheme: 'upto',
 *   network: 'eip155:8453',
 *   maxAmount: '1000000',  // $1.00 in USDC
 *   minAmount: '10000',    // $0.01 minimum
 *   asset: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
 *   payTo: '0x...',
 *   maxTimeoutSeconds: 300,
 *   extra: {
 *     unit: 'token',
 *     unitPrice: '100',
 *   },
 * };
 *
 * // Server settles for actual usage ($0.15)
 * const settlement: UptoSettlement = {
 *   settleAmount: '150000',
 *   usageDetails: {
 *     tokensGenerated: 1500,
 *     unitPrice: '100',
 *   },
 * };
 * ```
 */

import type { Network, PaymentRequirements } from "../";

/**
 * Extended payment requirements for the upto scheme.
 */
export interface UptoPaymentRequirements extends Omit<PaymentRequirements, "scheme" | "amount"> {
  /** Scheme identifier - always 'upto' */
  scheme: "upto";

  /** Network identifier (CAIP-2 format) */
  network: Network;

  /** Maximum amount the client authorizes (in smallest denomination) */
  maxAmount: string;

  /** Minimum settlement amount (prevents dust payments) */
  minAmount?: string;

  /** Asset contract address or identifier */
  asset: string;

  /** Recipient address */
  payTo: string;

  /** Maximum time in seconds before payment expires */
  maxTimeoutSeconds: number;

  /** Additional scheme-specific data */
  extra: UptoExtra;
}

/**
 * Extra fields specific to the upto scheme.
 */
export interface UptoExtra extends Record<string, unknown> {
  /** Billing unit (e.g., 'token', 'request', 'second', 'byte') */
  unit?: string;

  /** Price per unit in smallest denomination */
  unitPrice?: string;

  /** EIP-712 domain name (for EVM) */
  name?: string;

  /** EIP-712 domain version (for EVM) */
  version?: string;

  /** Router contract address (for EVM) */
  routerAddress?: string;
}

/**
 * Base payload structure for upto scheme.
 */
export interface UptoPayloadBase {
  /** Unique nonce to prevent replay attacks */
  nonce: string;
}

/**
 * EVM-specific upto payload using EIP-2612 Permit.
 */
export interface UptoEvmPayload extends UptoPayloadBase {
  /** EIP-2612 permit signature components */
  signature: {
    v: number;
    r: `0x${string}`;
    s: `0x${string}`;
  };

  /** Permit authorization parameters */
  authorization: {
    /** Token owner address */
    owner: `0x${string}`;

    /** Spender address (router contract) */
    spender: `0x${string}`;

    /** Maximum authorized value */
    value: string;

    /** Permit deadline (unix timestamp) */
    deadline: string;

    /** Permit nonce (from token contract) */
    nonce: number;
  };
}

/**
 * Alternative EVM payload with combined signature.
 */
export interface UptoEvmPayloadCompact extends UptoPayloadBase {
  /** Combined EIP-2612 permit signature */
  signature: `0x${string}`;

  /** Permit authorization parameters */
  authorization: {
    owner: `0x${string}`;
    spender: `0x${string}`;
    value: string;
    deadline: string;
    nonce: number;
  };
}

/**
 * Settlement request for upto scheme.
 */
export interface UptoSettlement {
  /** Actual amount to settle (must be <= maxAmount) */
  settleAmount: string;

  /** Optional usage details for auditing */
  usageDetails?: UptoUsageDetails;
}

/**
 * Usage details for settlement auditing.
 */
export interface UptoUsageDetails {
  /** Number of units consumed */
  unitsConsumed?: number;

  /** Price per unit used */
  unitPrice?: string;

  /** Type of unit */
  unitType?: string;

  /** Start timestamp of usage period */
  startTime?: number;

  /** End timestamp of usage period */
  endTime?: number;

  /** Additional metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Settlement response for upto scheme.
 */
export interface UptoSettlementResponse {
  /** Whether settlement was successful */
  success: boolean;

  /** Transaction hash (if on-chain) */
  transactionHash?: string;

  /** Actual amount settled */
  settledAmount: string;

  /** Maximum amount that was authorized */
  maxAmount: string;

  /** Block number (if on-chain) */
  blockNumber?: number;

  /** Gas used (if on-chain) */
  gasUsed?: string;

  /** Error message if failed */
  error?: string;
}

/**
 * Validation result for upto payment.
 */
export interface UptoValidationResult {
  /** Whether the payment is valid */
  isValid: boolean;

  /** Reason if invalid */
  invalidReason?: string;

  /** Validated maximum amount */
  validatedMaxAmount?: string;

  /** Payer address */
  payer?: string;

  /** Expiration timestamp */
  expiresAt?: number;
}

/**
 * Type guard for UptoPaymentRequirements.
 */
export function isUptoPaymentRequirements(
  requirements: unknown,
): requirements is UptoPaymentRequirements {
  if (typeof requirements !== "object" || requirements === null) return false;
  const req = requirements as Record<string, unknown>;
  return req.scheme === "upto" && "maxAmount" in req;
}

/**
 * Type guard for UptoEvmPayload.
 */
export function isUptoEvmPayload(
  payload: unknown,
): payload is UptoEvmPayload {
  if (typeof payload !== "object" || payload === null) return false;
  const p = payload as Record<string, unknown>;
  return (
    "signature" in p &&
    "authorization" in p &&
    typeof p.authorization === "object" &&
    p.authorization !== null &&
    "owner" in (p.authorization as Record<string, unknown>) &&
    "spender" in (p.authorization as Record<string, unknown>) &&
    "value" in (p.authorization as Record<string, unknown>) &&
    "deadline" in (p.authorization as Record<string, unknown>)
  );
}

/**
 * Constants for upto scheme.
 */
export const UPTO_SCHEME = "upto" as const;

export const UPTO_DEFAULTS = {
  /** Default minimum settlement amount (prevents dust) */
  MIN_AMOUNT: "1000",

  /** Default maximum timeout in seconds (5 minutes) */
  MAX_TIMEOUT_SECONDS: 300,

  /** Supported billing units */
  UNITS: ["token", "request", "second", "minute", "byte", "kb", "mb"] as const,
} as const;

export type UptoUnit = (typeof UPTO_DEFAULTS.UNITS)[number];

File: sdks/go/mechanisms/upto/upto.go
// Package upto provides shared types and constants for the Up-To payment scheme.
//
// The upto scheme authorizes transfer of up to a maximum amount, enabling
// usage-based billing where the final settlement amount is determined by
// actual usage. This is useful for AI inference billing, metered API access,
// streaming services, and data transfer billing.
//
// Chain-specific implementations live in their respective mechanism packages
// (e.g., mechanisms/evm/upto, mechanisms/svm/upto).
package upto

// Scheme is the identifier for the upto payment scheme.
const Scheme = "upto"

// DefaultMinAmount is the default minimum settlement amount (prevents dust).
const DefaultMinAmount = "1000"

// DefaultMaxTimeoutSeconds is the default maximum timeout in seconds.
const DefaultMaxTimeoutSeconds = 300

// SupportedUnits lists the valid billing unit types.
var SupportedUnits = []string{
	"token",
	"request",
	"second",
	"minute",
	"byte",
	"kb",
	"mb",
}

// IsValidUnit checks if the given unit is a supported billing unit.
func IsValidUnit(unit string) bool {
	for _, u := range SupportedUnits {
		if u == unit {
			return true
		}
	}
	return false
}

// Extra contains scheme-specific extra fields for upto payment requirements.
type Extra struct {
	// Unit is the billing unit (e.g., "token", "request", "second", "byte")
	Unit string `json:"unit,omitempty"`

	// UnitPrice is the price per unit in smallest denomination
	UnitPrice string `json:"unitPrice,omitempty"`

	// Name is the EIP-712 domain name (for EVM)
	Name string `json:"name,omitempty"`

	// Version is the EIP-712 domain version (for EVM)
	Version string `json:"version,omitempty"`

	// RouterAddress is the upto router contract address (for EVM)
	RouterAddress string `json:"routerAddress,omitempty"`
}

// PaymentRequirements represents payment requirements for the upto scheme.
type PaymentRequirements struct {
	// Scheme is always "upto"
	Scheme string `json:"scheme"`

	// Network is the network identifier in CAIP-2 format
	Network string `json:"network"`

	// MaxAmount is the maximum amount the client authorizes (in smallest denomination)
	MaxAmount string `json:"maxAmount"`

	// MinAmount is the minimum settlement amount (prevents dust payments)
	MinAmount string `json:"minAmount,omitempty"`

	// Asset is the token contract address or identifier
	Asset string `json:"asset"`

	// PayTo is the recipient address
	PayTo string `json:"payTo"`

	// MaxTimeoutSeconds is the maximum time before payment expires
	MaxTimeoutSeconds int `json:"maxTimeoutSeconds"`

	// Extra contains scheme-specific data
	Extra *Extra `json:"extra,omitempty"`
}

// UsageDetails contains usage information for settlement auditing.
type UsageDetails struct {
	// UnitsConsumed is the number of units consumed
	UnitsConsumed int `json:"unitsConsumed,omitempty"`

	// UnitPrice is the price per unit used
	UnitPrice string `json:"unitPrice,omitempty"`

	// UnitType is the type of unit
	UnitType string `json:"unitType,omitempty"`

	// StartTime is the start timestamp of the usage period
	StartTime int64 `json:"startTime,omitempty"`

	// EndTime is the end timestamp of the usage period
	EndTime int64 `json:"endTime,omitempty"`

	// Metadata contains additional tracking data
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// Settlement represents a settlement request for the upto scheme.
type Settlement struct {
	// SettleAmount is the actual amount to settle (must be <= maxAmount)
	SettleAmount string `json:"settleAmount"`

	// UsageDetails contains optional usage information
	UsageDetails *UsageDetails `json:"usageDetails,omitempty"`
}

// SettlementResponse represents a settlement response for the upto scheme.
type SettlementResponse struct {
	// Success indicates whether settlement was successful
	Success bool `json:"success"`

	// TransactionHash is the on-chain transaction hash (if applicable)
	TransactionHash string `json:"transactionHash,omitempty"`

	// SettledAmount is the actual amount settled
	SettledAmount string `json:"settledAmount"`

	// MaxAmount is the maximum amount that was authorized
	MaxAmount string `json:"maxAmount"`

	// BlockNumber is the block number (if on-chain)
	BlockNumber *int64 `json:"blockNumber,omitempty"`

	// GasUsed is the gas used (if on-chain)
	GasUsed string `json:"gasUsed,omitempty"`

	// Error is the error message if settlement failed
	Error string `json:"error,omitempty"`
}

// ValidationResult represents a validation result for an upto payment.
type ValidationResult struct {
	// IsValid indicates whether the payment is valid
	IsValid bool `json:"isValid"`

	// InvalidReason is the reason if invalid
	InvalidReason string `json:"invalidReason,omitempty"`

	// ValidatedMaxAmount is the verified maximum amount
	ValidatedMaxAmount string `json:"validatedMaxAmount,omitempty"`

	// Payer is the payer address
	Payer string `json:"payer,omitempty"`

	// ExpiresAt is the expiration timestamp
	ExpiresAt *int64 `json:"expiresAt,omitempty"`
}

// NewPaymentRequirements creates a new PaymentRequirements with default values.
func NewPaymentRequirements(network, maxAmount, asset, payTo string) *PaymentRequirements {
	return &PaymentRequirements{
		Scheme:            Scheme,
		Network:           network,
		MaxAmount:         maxAmount,
		MinAmount:         DefaultMinAmount,
		Asset:             asset,
		PayTo:             payTo,
		MaxTimeoutSeconds: DefaultMaxTimeoutSeconds,
	}
}

// NewSettlement creates a new Settlement.
func NewSettlement(settleAmount string) *Settlement {
	return &Settlement{
		SettleAmount: settleAmount,
	}
}

// NewSettlementWithUsage creates a new Settlement with usage details.
func NewSettlementWithUsage(settleAmount string, unitsConsumed int, unitPrice, unitType string) *Settlement {
	return &Settlement{
		SettleAmount: settleAmount,
		UsageDetails: &UsageDetails{
			UnitsConsumed: unitsConsumed,
			UnitPrice:     unitPrice,
			UnitType:      unitType,
		},
	}
}

// SuccessResponse creates a successful settlement response.
func SuccessResponse(settledAmount, maxAmount, txHash string) *SettlementResponse {
	return &SettlementResponse{
		Success:         true,
		SettledAmount:   settledAmount,
		MaxAmount:       maxAmount,
		TransactionHash: txHash,
	}
}

// FailureResponse creates a failed settlement response.
func FailureResponse(maxAmount, errMsg string) *SettlementResponse {
	return &SettlementResponse{
		Success:       false,
		SettledAmount: "0",
		MaxAmount:     maxAmount,
		Error:         errMsg,
	}
}

// Valid creates a valid ValidationResult.
func Valid(validatedMaxAmount, payer string, expiresAt int64) *ValidationResult {
	return &ValidationResult{
		IsValid:            true,
		ValidatedMaxAmount: validatedMaxAmount,
		Payer:              payer,
		ExpiresAt:          &expiresAt,
	}
}

// Invalid creates an invalid ValidationResult.
func Invalid(reason string) *ValidationResult {
	return &ValidationResult{
		IsValid:       false,
		InvalidReason: reason,
	}
}

// IsUptoRequirements checks if the given data represents upto payment requirements.
func IsUptoRequirements(data map[string]interface{}) bool {
	scheme, ok := data["scheme"].(string)
	if !ok || scheme != Scheme {
		return false
	}
	_, hasMaxAmount := data["maxAmount"]
	return hasMaxAmount
}
